File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit_objc/Protocols/FLTPaymentQueueProtocol.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "../include/in_app_purchase_storekit_objc/FLTPaymentQueueProtocol.h"

@interface DefaultPaymentQueue ()
/// The wrapped SKPaymentQueue
@property(nonatomic, strong) SKPaymentQueue *queue;
@end

@implementation DefaultPaymentQueue

@synthesize storefront;
@synthesize delegate;
@synthesize transactions;

- (instancetype)initWithQueue:(SKPaymentQueue *)queue {
  self = [super init];
  if (self) {
    _queue = queue;
  }
  return self;
}

- (void)addPayment:(SKPayment *_Nonnull)payment {
  [self.queue addPayment:payment];
}

- (void)finishTransaction:(nonnull SKPaymentTransaction *)transaction {
  [self.queue finishTransaction:transaction];
}

- (void)addTransactionObserver:(nonnull id<SKPaymentTransactionObserver>)observer {
  [self.queue addTransactionObserver:observer];
}

- (void)restoreCompletedTransactions {
  [self.queue restoreCompletedTransactions];
}

- (void)restoreCompletedTransactionsWithApplicationUsername:(nullable NSString *)username {
  [self.queue restoreCompletedTransactionsWithApplicationUsername:username];
}

- (id<SKPaymentQueueDelegate>)delegate API_AVAILABLE(ios(13.0), macos(10.15), watchos(6.2)) {
  return self.queue.delegate;
}

- (NSArray<SKPaymentTransaction *> *)transactions API_AVAILABLE(ios(3.0), macos(10.7),
                                                                watchos(6.2)) {
  return self.queue.transactions;
}

- (SKStorefront *)storefront API_AVAILABLE(ios(13.0)) {
  return self.queue.storefront;
}

#if TARGET_OS_IOS
- (void)presentCodeRedemptionSheet API_AVAILABLE(ios(14.0))API_UNAVAILABLE(tvos, macos, watchos) {
  [self.queue presentCodeRedemptionSheet];
}
#endif

#if TARGET_OS_IOS
- (void)showPriceConsentIfNeeded API_AVAILABLE(ios(13.4))API_UNAVAILABLE(tvos, macos, watchos) {
  [self.queue showPriceConsentIfNeeded];
}
#endif

@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/Protocols/FLTPaymentQueueProtocol.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import StoreKit

/// A protocol that wraps SKPaymentQueue
@objc public protocol FLTPaymentQueueProtocol: NSObjectProtocol {
  /// An object containing the location and unique identifier of an Apple App Store storefront.
  @available(iOS 13.0, *)
  var storefront: SKStorefront? { get set }

  /// A list of SKPaymentTransactions, which each represents a single transaction
  var transactions: [SKPaymentTransaction] { get set }

  /// An object that provides information needed to complete transactions.
  @available(iOS 13.0, macOS 10.15, watchOS 6.2, *)
  weak var delegate: SKPaymentQueueDelegate? { get set }

  /// Remove a finished (i.e. failed or completed) transaction from the queue.  Attempting to finish a
  /// purchasing transaction will throw an exception.
  func finish(_ transaction: SKPaymentTransaction)

  /// Observers are not retained.  The transactions array will only be synchronized with the server
  /// while the queue has observers.  This may require that the user authenticate.
  @objc(addTransactionObserver:)
  func add(_ observer: SKPaymentTransactionObserver)

  /// Add a payment to the server queue.  The payment is copied to add an SKPaymentTransaction to the
  /// transactions array.  The same payment can be added multiple times to create multiple
  /// transactions.
  @objc(addPayment:)
  func add(_ payment: SKPayment)

  /// Will add completed transactions for the current user back to the queue to be re-completed.
  func restoreCompletedTransactions()

  /// Will add completed transactions for the current user back to the queue to be re-completed. This
  /// version requires an identifier to the user's account.
  func restoreCompletedTransactions(withApplicationUsername username: String?)

  #if os(iOS)
    /// Call this method to have StoreKit present a sheet enabling the user to redeem codes provided by
    /// your app. Only for iOS.
    @available(iOS 14.0, *)
    func presentCodeRedemptionSheet()

    /// If StoreKit has called your SKPaymentQueueDelegate's "paymentQueueShouldShowPriceConsent:"
    /// method and you returned NO, you can use this method to show the price consent UI at a later time
    /// that is more appropriate for your app. If there is no pending price consent, this method will do
    /// nothing.
    @available(iOS 13.4, *)
    func showPriceConsentIfNeeded()
  #endif
}

/// The default PaymentQueue that wraps SKPaymentQueue
public class DefaultPaymentQueue: NSObject, FLTPaymentQueueProtocol {
  /// The wrapped SKPaymentQueue
  private let queue: SKPaymentQueue

  private weak var _delegate: SKPaymentQueueDelegate?

  /// Initialize this wrapper with an SKPaymentQueue
  public init(queue: SKPaymentQueue) {
    self.queue = queue
  }

  @objc(addPayment:)
  public func add(_ payment: SKPayment) {
    queue.add(payment)
  }

  public func finish(_ transaction: SKPaymentTransaction) {
    queue.finishTransaction(transaction)
  }

  @objc(addTransactionObserver:)
  public func add(_ observer: SKPaymentTransactionObserver) {
    queue.add(observer)
  }

  public func restoreCompletedTransactions() {
    queue.restoreCompletedTransactions()
  }

  public func restoreCompletedTransactions(withApplicationUsername username: String?) {
    queue.restoreCompletedTransactions(withApplicationUsername: username)
  }

  @available(iOS 13.0, macOS 10.15, watchOS 6.2, *)
  public var delegate: SKPaymentQueueDelegate? {
    get { queue.delegate }
    set { _delegate = newValue }
  }

  public var transactions: [SKPaymentTransaction] {
    get { queue.transactions }
    set {}
  }

  @available(iOS 13.0, *)
  public var storefront: SKStorefront? {
    get { queue.storefront }
    set {}
  }

  #if os(iOS)
    @available(iOS 14.0, *)
    public func presentCodeRedemptionSheet() {
      queue.presentCodeRedemptionSheet()
    }

    @available(iOS 13.4, *)
    public func showPriceConsentIfNeeded() {
      queue.showPriceConsentIfNeeded()
    }
  #endif
}
